### DISCUSSION TITLE
How to make the websocket connection end-to-end encrypted?

### DISCUSSION THREAD

**Discussion by ghost** [Totodore/socketioxide#308] [Q&A] [answered]
I'm working on a project with websockets that requires alot of security measures and I would like to implement E2EE but I'm not sure how. The first idea I had was using TLS but that is beyond my knowledge.

Could I get some help with this?

**Comment by Totodore** [comment, marked as answer]
Hello, the TLS encryption is handled by the underlying http lib.
Check this example:
https://github.com/Totodore/socketioxide/tree/main/examples/axum-echo-tls

**Reply by ghost** [reply]
That looks very interesting thank you, could you explain the differences between message and message-with-ack?
Also a bit about how this would work in a realistic situation?

**Reply by Totodore** [reply]
I suppose you mean [`emit`](https://docs.rs/socketioxide/latest/socketioxide/socket/struct.Socket.html#method.emit) and [`emit_with_ack`](https://docs.rs/socketioxide/latest/socketioxide/socket/struct.Socket.html#method.emit_with_ack)? With `emit_with_ack` you can expect an acknowledgement (see official [socket.io doc](https://socket.io/docs/v4/tutorial/api-overview#acknowledgements)) by awaiting on the `Ok` value returned by the method.

**Reply by ghost** [reply]
The docs were super helpful for understanding acknoledgements. Thank you very much!
